This notebook will train and validate a number of Machine Learning algorithms to classify reports as either positive or negative for fluid collection.

The first part of this will read in the data and transform it from free text into sparse vectors. The second part will train the algorithms and evaluate their performance.

# Data Processing

Right now the data is stored in a sqlite database. There are two main columns:
- ``text``: the free text of the radiology report
- ``doc_class``: a 1 if there is a fluid collection, 0 if there isn't.

They need to be represented as sparse vectors. We'll read them in, preprocess them, and convert them using ``sklearn``.

In [1]:
import os
import pandas as pd
import sqlite3 as sqlite

from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import cross_val_score
from sklearn.metrics import f1_score,accuracy_score, precision_score, recall_score, classification_report
from sklearn.naive_bayes import MultinomialNB
from sklearn.linear_model import LogisticRegression
from sklearn.svm import LinearSVC
from sklearn.utils import shuffle
import pickle

In [2]:
DATADIR = '../stats_data'
DB = os.path.join(DATADIR, 'Reference Standard', 'radiology_reports.sqlite')
os.path.exists(DB)

True

In [3]:
conn = sqlite.connect(DB)

# Training data
train_df = pd.read_sql("SELECT * FROM training_notes;", conn)
#Testing data
test_df = pd.read_sql("SELECT * FROM testing_notes;", conn)
conn.close()

In [4]:
train_df.head()

rowid                      name  \
0      1  No_10792_131562_05-29-20   
1      2  No_11050_126785_11-03-33   
2      3  No_11879_166554_06-22-37   
3      4  No_11879_166554_06-23-37   
4      5  No_11879_166554_07-02-37   

                                                text  \
0   \n CT ABDOMEN W/CONTRAST; CT PELVIS W/CONTRAS...   
1   \n CT CHEST W/CONTRAST; CT ABDOMEN W/CONTRAST...   
2   \n CTA CHEST W&W/O C &RECONS; CT 100CC NON IO...   
3   \n CT ABDOMEN W/O CONTRAST; CT PELVIS W/O CON...   
4   \n CT CHEST W/O CONTRAST \n ~ Reason: r/o ste...   

                                        referenceXML  doc_class  subject  \
0  <?xml version="1.0" encoding="UTF-8"?>\n<annot...          0       32   
1  <?xml version="1.0" encoding="UTF-8"?>\n<annot...          0       34   
2  <?xml version="1.0" encoding="UTF-8"?>\n<annot...          0       35   
3  <?xml version="1.0" encoding="UTF-8"?>\n<annot...          0       35   
4  <?xml version="1.0" encoding="UTF-8"?>\n<annot...          0       35   

   HADM_ID CHARTDATE  
0   131562  05-29-20  
1   126785  11-03-33  
2   166554  06-22-37  
3   166554  06-23-37  
4   166554  07-02-37

In [5]:
test_df.shape

(100, 8)

The `TfidfVectorizer` and `CountVectorizer` classes transform raw texts into matrices where the rows represent reports and the columns represents terms that are present in that report. `Tfidf` uses *Term-Freqeuncy Inverse-Document Frequency* weighting, while `Count` uses raw counts.

We'll use the vectorizer to preprocess the text, as well. We'll make the reports lowercase, remove all stopwords, and extracts ngrams from 1-3. We'll also set a minimum document threshold, stating that an ngram must appear in at least 10% of the documents to be included. This will help cut down on a lot of noise.

In [6]:
vectorizer = TfidfVectorizer(min_df=0.1, lowercase=True, stop_words='english',
                                    ngram_range=(1, 3))

X_train = list(train_df.text)
y_train = list(train_df.doc_class)

# Fit the vectorizer and transform the training notes into a vector
X_train = vectorizer.fit_transform(X_train)


# Now use this fitted vectorizer to transform the test data
X_test = list(test_df.text)
y_test = list(test_df.doc_class)
X_test = vectorizer.transform(X_test)

In [7]:
X_train.todense()

matrix([[0.        , 0.04251338, 0.04533661, ..., 0.        , 0.06027539,
         0.        ],
        [0.        , 0.        , 0.        , ..., 0.        , 0.02942781,
         0.        ],
        [0.08356349, 0.        , 0.        , ..., 0.        , 0.04091356,
         0.        ],
        ...,
        [0.05500347, 0.        , 0.        , ..., 0.        , 0.        ,
         0.        ],
        [0.        , 0.        , 0.        , ..., 0.04392229, 0.02344006,
         0.        ],
        [0.        , 0.        , 0.        , ..., 0.06829954, 0.0364495 ,
         0.        ]])

In [8]:
# Shuffle the training data
X_train, y_train = shuffle(X_train, y_train)

In [9]:
print(X_train.shape)
print(X_test.shape)

(545, 682)
(100, 682)


Our vocabulary consists of 682 ngrams. Our data has 545 training notes and 100 testing notes.

Now we can start training.

# Machine Learning

We'll consider these algorithms as document classifiers:
- Logistic Regression
- Random Forest
- Naive Bayes
- Linear SVM

For each of these, we'll also consider a number of hyper-parameters to try and find the optimal performance.

In [10]:
def train_and_evaluate_model(X_train, y_train, X_test, y_test, clf, model_name, results):
    print(model_name)
    clf.fit(X_train, y_train)
    y_pred_train = clf.predict(X_train)
    f1, p, r = (f1_score(y_train, y_pred_train,average='binary'),
               precision_score(y_train, y_pred_train, average='binary'),
               recall_score(y_train, y_pred_train,average='binary'))
    print("Train:")
    print(f1, p, r)
    results = results.append({'test/train':'train','name':model_name,
                              'F1':f1,'Precision':p,'Recall':r,
                              "Classifier": clf},ignore_index=True)

    # Now evaluate on testing data
    y_pred_test = clf.predict(X_test)
    f1, p, r = (f1_score(y_test, y_pred_test, average='binary'),
               precision_score(y_test, y_pred_test, average='binary'),
               recall_score(y_test, y_pred_test, average='binary'))
    print("Test:")
    print(f1, p, r)
    results = results.append({'test/train':'test', 
                              'name':model_name,
                              'F1':f1, 'Precision':p, 'Recall':r,
                             "Classifier": clf},
                             ignore_index=True)
    return results

In [11]:
# This dataframe will keep track of all of our scores
results = pd.DataFrame(columns=('test/train','name', 'F1',"Precision","Recall", "Classifier"))


### Logistic Regression

In [12]:
clf = LogisticRegression()
penalties = ['l1', 'l2']
Cs = [0.01, 0.1, 0.5, 1.0]

for penalty in penalties:
    for c in Cs:
        model_name = "Logistic Regression: penalty={}, C={}".format(penalty, c)
        clf = LogisticRegression(penalty=penalty, C=c)
        
        results = train_and_evaluate_model(X_train, y_train, X_test, y_test, clf, model_name, results)
print(results.tail())

Logistic Regression: penalty=l1, C=0.01
Train:
0.0 0.0 0.0
Test:
0.0 0.0 0.0
Logistic Regression: penalty=l1, C=0.1
Train:
0.07377049180327869 0.9 0.038461538461538464
Test:
0.17647058823529413 1.0 0.0967741935483871
Logistic Regression: penalty=l1, C=0.5
Train:
0.7358490566037735 0.8210526315789474 0.6666666666666666
Test:
0.6666666666666667 0.782608695652174 0.5806451612903226
Logistic Regression: penalty=l1, C=1.0
Train:
0.7788018433179723 0.845 0.7222222222222222
Test:
0.6666666666666667 0.782608695652174 0.5806451612903226
Logistic Regression: penalty=l2, C=0.01
Train:
0.0 0.0 0.0
Test:
0.0 0.0 0.0
Logistic Regression: penalty=l2, C=0.1
Train:
0.6144927536231884 0.954954954954955 0.452991452991453
Test:
0.5454545454545454 0.9230769230769231 0.3870967741935484
Logistic Regression: penalty=l2, C=0.5


/Users/alec/anaconda/envs/i2b2/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)
/Users/alec/anaconda/envs/i2b2/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)


Train:
0.8037825059101654 0.8994708994708994 0.7264957264957265
Test:
0.7777777777777777 0.9130434782608695 0.6774193548387096
Logistic Regression: penalty=l2, C=1.0
Train:
0.8375286041189931 0.9014778325123153 0.782051282051282
Test:
0.7857142857142856 0.88 0.7096774193548387
   test/train                                    name        F1  Precision  \
11       test  Logistic Regression: penalty=l2, C=0.1  0.545455   0.923077   
12      train  Logistic Regression: penalty=l2, C=0.5  0.803783   0.899471   
13       test  Logistic Regression: penalty=l2, C=0.5  0.777778   0.913043   
14      train  Logistic Regression: penalty=l2, C=1.0  0.837529   0.901478   
15       test  Logistic Regression: penalty=l2, C=1.0  0.785714   0.880000   

      Recall                                         Classifier  
11  0.387097  LogisticRegression(C=0.1, class_weight=None, d...  
12  0.726496  LogisticRegression(C=0.5, class_weight=None, d...  
13  0.677419  LogisticRegression(C=0.5, class_weight=No

### Random Forest

In [13]:
n_estimators = [200,500,1000]
max_features = ['sqrt',None]
max_depth = [50,None]
for feat in max_features:
    for d in max_depth:
        for n in n_estimators:
            # Train and evaluate on training data
            model_name = 'Random Forest: n={n}, feat={f}, depth={d}'.format(n=n,f=feat,d=d)
            print(model_name)
            forest = RandomForestClassifier(n_estimators=n,max_depth=d,max_features=feat, n_jobs=-1)
            results = train_and_evaluate_model(X_train, y_train, X_test, y_test, forest, model_name, results)
        
results

Random Forest: n=200, feat=sqrt, depth=50
Random Forest: n=200, feat=sqrt, depth=50
Train:
1.0 1.0 1.0
Test:
0.7241379310344828 0.7777777777777778 0.6774193548387096
Random Forest: n=500, feat=sqrt, depth=50
Random Forest: n=500, feat=sqrt, depth=50
Train:
1.0 1.0 1.0
Test:
0.7586206896551724 0.8148148148148148 0.7096774193548387
Random Forest: n=1000, feat=sqrt, depth=50
Random Forest: n=1000, feat=sqrt, depth=50
Train:
1.0 1.0 1.0
Test:
0.7368421052631579 0.8076923076923077 0.6774193548387096
Random Forest: n=200, feat=sqrt, depth=None
Random Forest: n=200, feat=sqrt, depth=None
Train:
1.0 1.0 1.0
Test:
0.7368421052631579 0.8076923076923077 0.6774193548387096
Random Forest: n=500, feat=sqrt, depth=None
Random Forest: n=500, feat=sqrt, depth=None
Train:
1.0 1.0 1.0
Test:
0.7586206896551724 0.8148148148148148 0.7096774193548387
Random Forest: n=1000, feat=sqrt, depth=None
Random Forest: n=1000, feat=sqrt, depth=None
Train:
1.0 1.0 1.0
Test:
0.7796610169491526 0.8214285714285714 0.74193

test/train                                          name        F1  \
0       train       Logistic Regression: penalty=l1, C=0.01  0.000000   
1        test       Logistic Regression: penalty=l1, C=0.01  0.000000   
2       train        Logistic Regression: penalty=l1, C=0.1  0.073770   
3        test        Logistic Regression: penalty=l1, C=0.1  0.176471   
4       train        Logistic Regression: penalty=l1, C=0.5  0.735849   
5        test        Logistic Regression: penalty=l1, C=0.5  0.666667   
6       train        Logistic Regression: penalty=l1, C=1.0  0.778802   
7        test        Logistic Regression: penalty=l1, C=1.0  0.666667   
8       train       Logistic Regression: penalty=l2, C=0.01  0.000000   
9        test       Logistic Regression: penalty=l2, C=0.01  0.000000   
10      train        Logistic Regression: penalty=l2, C=0.1  0.614493   
11       test        Logistic Regression: penalty=l2, C=0.1  0.545455   
12      train        Logistic Regression: penalty=l2, C=0.5  0.803783   
13       test        Logistic Regression: penalty=l2, C=0.5  0.777778   
14      train        Logistic Regression: penalty=l2, C=1.0  0.837529   
15       test        Logistic Regression: penalty=l2, C=1.0  0.785714   
16      train     Random Forest: n=200, feat=sqrt, depth=50  1.000000   
17       test     Random Forest: n=200, feat=sqrt, depth=50  0.724138   
18      train     Random Forest: n=500, feat=sqrt, depth=50  1.000000   
19       test     Random Forest: n=500, feat=sqrt, depth=50  0.758621   
20      train    Random Forest: n=1000, feat=sqrt, depth=50  1.000000   
21       test    Random Forest: n=1000, feat=sqrt, depth=50  0.736842   
22      train   Random Forest: n=200, feat=sqrt, depth=None  1.000000   
23       test   Random Forest: n=200, feat=sqrt, depth=None  0.736842   
24      train   Random Forest: n=500, feat=sqrt, depth=None  1.000000   
25       test   Random Forest: n=500, feat=sqrt, depth=None  0.758621   
26      train  Random Forest: n=1000, feat=sqrt, depth=None  1.000000   
27       test  Random Forest: n=1000, feat=sqrt, depth=None  0.779661   
28      train     Random Forest: n=200, feat=None, depth=50  1.000000   
29       test     Random Forest: n=200, feat=None, depth=50  0.727273   
30      train     Random Forest: n=500, feat=None, depth=50  1.000000   
31       test     Random Forest: n=500, feat=None, depth=50  0.746269   
32      train    Random Forest: n=1000, feat=None, depth=50  1.000000   
33       test    Random Forest: n=1000, feat=None, depth=50  0.746269   
34      train   Random Forest: n=200, feat=None, depth=None  1.000000   
35       test   Random Forest: n=200, feat=None, depth=None  0.738462   
36      train   Random Forest: n=500, feat=None, depth=None  1.000000   
37       test   Random Forest: n=500, feat=None, depth=None  0.727273   
38      train  Random Forest: n=1000, feat=None, depth=None  1.000000   
39       test  Random Forest: n=1000, feat=None, depth=None  0.757576   

    Precision    Recall                                         Classifier  
0    0.000000  0.000000  LogisticRegression(C=0.01, class_weight=None, ...  
1    0.000000  0.000000  LogisticRegression(C=0.01, class_weight=None, ...  
2    0.900000  0.038462  LogisticRegression(C=0.1, class_weight=None, d...  
3    1.000000  0.096774  LogisticRegression(C=0.1, class_weight=None, d...  
4    0.821053  0.666667  LogisticRegression(C=0.5, class_weight=None, d...  
5    0.782609  0.580645  LogisticRegression(C=0.5, class_weight=None, d...  
6    0.845000  0.722222  LogisticRegression(C=1.0, class_weight=None, d...  
7    0.782609  0.580645  LogisticRegression(C=1.0, class_weight=None, d...  
8    0.000000  0.000000  LogisticRegression(C=0.01, class_weight=None, ...  
9    0.000000  0.000000  LogisticRegression(C=0.01, class_weight=None, ...  
10   0.954955  0.452991  LogisticRegression(C=0.1, class_weight=None, d...  
11   0.923077  0.387097  LogisticRegression(C=0.1, class_weight=None, d...  
12   0.8

### Naive Bayes

In [14]:
# There are no hyperparemters for this algorithm
mnb = MultinomialNB()
results = train_and_evaluate_model(X_train, y_train, X_test, y_test, mnb, "Naive Bayes", results)
results.tail()

Naive Bayes
Train:
0.7695749440715883 0.8075117370892019 0.7350427350427351
Test:
0.7796610169491526 0.8214285714285714 0.7419354838709677


test/train                                          name        F1  \
37       test   Random Forest: n=500, feat=None, depth=None  0.727273   
38      train  Random Forest: n=1000, feat=None, depth=None  1.000000   
39       test  Random Forest: n=1000, feat=None, depth=None  0.757576   
40      train                                   Naive Bayes  0.769575   
41       test                                   Naive Bayes  0.779661   

    Precision    Recall                                         Classifier  
37   0.685714  0.774194  (DecisionTreeClassifier(class_weight=None, cri...  
38   1.000000  1.000000  (DecisionTreeClassifier(class_weight=None, cri...  
39   0.714286  0.806452  (DecisionTreeClassifier(class_weight=None, cri...  
40   0.807512  0.735043  MultinomialNB(alpha=1.0, class_prior=None, fit...  
41   0.821429  0.741935  MultinomialNB(alpha=1.0, class_prior=None, fit...

### Linear SVM

In [15]:
slacks = [2, 1, 0.1, 0.05, 0.01]
for slack in slacks:
    model_name = "SVM: slack={slack}".format(slack=slack)
    svc_clf = LinearSVC(C=float(slack),loss='hinge')
    results = train_and_evaluate_model(X_train, y_train, X_test, y_test, mnb, model_name, results)
print(results.tail())

SVM: slack=2
Train:
0.7695749440715883 0.8075117370892019 0.7350427350427351
Test:
0.7796610169491526 0.8214285714285714 0.7419354838709677
SVM: slack=1
Train:
0.7695749440715883 0.8075117370892019 0.7350427350427351
Test:
0.7796610169491526 0.8214285714285714 0.7419354838709677
SVM: slack=0.1
Train:
0.7695749440715883 0.8075117370892019 0.7350427350427351
Test:
0.7796610169491526 0.8214285714285714 0.7419354838709677
SVM: slack=0.05
Train:
0.7695749440715883 0.8075117370892019 0.7350427350427351
Test:
0.7796610169491526 0.8214285714285714 0.7419354838709677
SVM: slack=0.01
Train:
0.7695749440715883 0.8075117370892019 0.7350427350427351
Test:
0.7796610169491526 0.8214285714285714 0.7419354838709677
   test/train             name        F1  Precision    Recall  \
47       test   SVM: slack=0.1  0.779661   0.821429  0.741935   
48      train  SVM: slack=0.05  0.769575   0.807512  0.735043   
49       test  SVM: slack=0.05  0.779661   0.821429  0.741935   
50      train  SVM: slack=0.01  

# Analysis

Now that we've evaluated a number of classifiers, we'll pick the best one, save its predictions, and do a further analysis on it.

In [16]:
results.head()

test/train                                     name        F1  Precision  \
0      train  Logistic Regression: penalty=l1, C=0.01  0.000000   0.000000   
1       test  Logistic Regression: penalty=l1, C=0.01  0.000000   0.000000   
2      train   Logistic Regression: penalty=l1, C=0.1  0.073770   0.900000   
3       test   Logistic Regression: penalty=l1, C=0.1  0.176471   1.000000   
4      train   Logistic Regression: penalty=l1, C=0.5  0.735849   0.821053   

     Recall                                         Classifier  
0  0.000000  LogisticRegression(C=0.01, class_weight=None, ...  
1  0.000000  LogisticRegression(C=0.01, class_weight=None, ...  
2  0.038462  LogisticRegression(C=0.1, class_weight=None, d...  
3  0.096774  LogisticRegression(C=0.1, class_weight=None, d...  
4  0.666667  LogisticRegression(C=0.5, class_weight=None, d...

In [28]:
# Filter to only look at test scores
# Then sort by F1
results[results['test/train'] == 'test'].sort_values(by='F1', ascending=False)

test/train                                          name        F1  \
15       test        Logistic Regression: penalty=l2, C=1.0  0.785714   
27       test  Random Forest: n=1000, feat=sqrt, depth=None  0.779661   
49       test                               SVM: slack=0.05  0.779661   
47       test                                SVM: slack=0.1  0.779661   
45       test                                  SVM: slack=1  0.779661   
43       test                                  SVM: slack=2  0.779661   
41       test                                   Naive Bayes  0.779661   
51       test                               SVM: slack=0.01  0.779661   
13       test        Logistic Regression: penalty=l2, C=0.5  0.777778   
25       test   Random Forest: n=500, feat=sqrt, depth=None  0.758621   
19       test     Random Forest: n=500, feat=sqrt, depth=50  0.758621   
39       test  Random Forest: n=1000, feat=None, depth=None  0.757576   
31       test     Random Forest: n=500, feat=None, depth=50  0.746269   
33       test    Random Forest: n=1000, feat=None, depth=50  0.746269   
35       test   Random Forest: n=200, feat=None, depth=None  0.738462   
21       test    Random Forest: n=1000, feat=sqrt, depth=50  0.736842   
23       test   Random Forest: n=200, feat=sqrt, depth=None  0.736842   
29       test     Random Forest: n=200, feat=None, depth=50  0.727273   
37       test   Random Forest: n=500, feat=None, depth=None  0.727273   
17       test     Random Forest: n=200, feat=sqrt, depth=50  0.724138   
7        test        Logistic Regression: penalty=l1, C=1.0  0.666667   
5        test        Logistic Regression: penalty=l1, C=0.5  0.666667   
11       test        Logistic Regression: penalty=l2, C=0.1  0.545455   
3        test        Logistic Regression: penalty=l1, C=0.1  0.176471   
9        test       Logistic Regression: penalty=l2, C=0.01  0.000000   
1        test       Logistic Regression: penalty=l1, C=0.01  0.000000   

    Precision    Recall                                         Classifier  
15   0.880000  0.709677  LogisticRegression(C=1.0, class_weight=None, d...  
27   0.821429  0.741935  (DecisionTreeClassifier(class_weight=None, cri...  
49   0.821429  0.741935  MultinomialNB(alpha=1.0, class_prior=None, fit...  
47   0.821429  0.741935  MultinomialNB(alpha=1.0, class_prior=None, fit...  
45   0.821429  0.741935  MultinomialNB(alpha=1.0, class_prior=None, fit...  
43   0.821429  0.741935  MultinomialNB(alpha=1.0, class_prior=None, fit...  
41   0.821429  0.741935  MultinomialNB(alpha=1.0, class_prior=None, fit...  
51   0.821429  0.741935  MultinomialNB(alpha=1.0, class_prior=None, fit...  
13   0.913043  0.677419  LogisticRegression(C=0.5, class_weight=None, d...  
25   0.814815  0.709677  (DecisionTreeClassifier(class_weight=None, cri...  
19   0.814815  0.709677  (DecisionTreeClassifier(class_weight=None, cri...  
39   0.714286  0.806452  (DecisionTreeClassifier(class_weight=None, cri...  
31   0.694444  0.806452  (DecisionTreeClassifier(class_weight=None, cri...  
33   0.694444  0.806452  (DecisionTreeClassifier(class_weight=None, cri...  
35   0.705882  0.774194  (DecisionTreeClassifier(class_weight=None, cri...  
21   0.807692  0.677419  (DecisionTreeClassifier(class_weight=None, cri...  
23   0.807692  0.677419  (DecisionTreeClassifier(class_weight=None, cri...  
29   0.685714  0.774194  (DecisionTreeClassifier(class_weight=None, cri...  
37   0.685714  0.774194  (DecisionTreeClassifier(class_weight=None, cri...  
17   0.777778  0.677419  (DecisionTreeClassifier(class_weight=None, cri...  
7    0.782609  0.580645  LogisticRegression(C=1.0, class_weight=None, d...  
5    0.782609  0.580645  LogisticRegression(C=0.5, class_weight=None, d...  
11   0.923077  0.387097  LogisticRegression(C=0.1, class_weight=None, d...  
3    1.000000  0.096774  LogisticRegression(C=0.1, class_weight=None, d...  
9    0.000000  0.000000  LogisticRegression(C=0.01, class_weight=None, ...  
1    0.000000  0.000000  Logi

In [18]:
results.iloc[23]

test/train                                                 test
name                Random Forest: n=200, feat=sqrt, depth=None
F1                                                     0.736842
Precision                                              0.807692
Recall                                                 0.677419
Classifier    (DecisionTreeClassifier(class_weight=None, cri...
Name: 23, dtype: object

Looks like our highest-performing model is a **Logistic Regression Classifier** with l2 regularization and a regularization parameter of C=1.0. Let's now retrain that model and save its predictions:

In [29]:
clf = results.iloc[15].Classifier
print(clf)
y_pred_test = clf.predict(X_test)
y_pred_test

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)


array([0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1,
       0, 0, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0,
       1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1,
       0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1])

In [30]:
print(classification_report(y_test, y_pred_test))

             precision    recall  f1-score   support

          0       0.88      0.96      0.92        69
          1       0.88      0.71      0.79        31

avg / total       0.88      0.88      0.88       100



In [31]:
# Now we'll save the dataframe that also has predictions
test_df.to_pickle('../stats_data/test_data_with_preds.pkl')

In [37]:
# Let's save just the predictions
test_df = pd.read_pickle('../stats_data/test_data_with_preds.pkl')
test_df['pred'] = y_pred_test
test_df.head()

rowid                      name  \
0      1   No_1007_141227_06-18-95   
1      2  No_12344_140694_08-13-21   
2      3  No_14176_126791_10-19-39   
3      4  No_15847_121459_06-07-77   
4      5  No_15847_121459_06-16-77   

                                                text  \
0   \n CT CHEST W/CONTRAST \n ~ Reason: assess de...   
1   \n CTA CHEST W&W/O C&RECONS, NON-CORONARY; CT...   
2   \n CTA CHEST W&W/O C&RECONS, NON-CORONARY; CT...   
3   \n CTA ABD W&W/O C & RECONS; CTA PELVIS W&W/O...   
4   \n CT ABDOMEN W/O CONTRAST; CT PELVIS W/O CON...   

                                        referenceXML  doc_class  subject  \
0  <?xml version="1.0" encoding="UTF-8"?>\n<annot...          0        5   
1  <?xml version="1.0" encoding="UTF-8"?>\n<annot...          0       37   
2  <?xml version="1.0" encoding="UTF-8"?>\n<annot...          0       41   
3  <?xml version="1.0" encoding="UTF-8"?>\n<annot...          0       45   
4  <?xml version="1.0" encoding="UTF-8"?>\n<annot...          0       45   

   HADM_ID CHARTDATE  pred  
0   141227  06-18-95     0  
1   140694  08-13-21     0  
2   126791  10-19-39     0  
3   121459  06-07-77     0  
4   121459  06-16-77     0

In [39]:
# Save the classifier
with open('../stats_data/lin_reg.pkl', 'wb') as f:
    pickle.dump(clf, f)
    
# Save the vectorizer and testing data
with open('../stats_data/test_data.pkl', 'wb') as f:
    pickle.dump((X_test, y_test, vectorizer), f)

In [40]:
out = test_df[['name', 'doc_class', 'pred']]
out.head()

name  doc_class  pred
0   No_1007_141227_06-18-95          0     0
1  No_12344_140694_08-13-21          0     0
2  No_14176_126791_10-19-39          0     0
3  No_15847_121459_06-07-77          0     0
4  No_15847_121459_06-16-77          0     0

In [41]:
out.to_csv('../stats_data/test_preds.txt', sep='\t')

# Conclusion
Our best performing classifier was a Logistic Regression. We ended up with an F1 of 0.79 for positive instances. This is a reasonably good score. In our final notebook we'll test our predictions for statistical significance and see if this classifier is significantly better than others

# Up Next
[Analysis](Analysis.ipynb)